### PR TITLE
ENH: generate Dockerfiles with neurodocker + migrate to CircleCI 2.0

### LINKED ISSUE #2144: ENH proposal: use neurodocker to generate and minimize containers for testing
@oesteban @effigies @chrisfilo @djarecka 

i would like to change nipype's container generation and testing using the following approach:

1. generate dockerfiles using neurodocker
2. check if dockerfile hash has changed or force update is set (need to determine this from PR)
3. if force update, regenerate dockerfile
4. build docker image if needed or reload minimized image
5. run all tests via neurodocker reprozip (in parallel)
6. combine all reprozip packs into a single pack
7. generate minimized docker image using reprounzip and cache this image

if this sounds good, perhaps @kaczmarj can work on implementing this.

### PR DESCRIPTION
**Pull request by kaczmarj**
Fixes #2144.

Changes proposed in this pull request:

Addresses (or will address) these items from the original issue:
> 1. generate dockerfiles using neurodocker
> 2. check if dockerfile hash has changed or force update is set (need to determine this from PR)
> 3. if force update, regenerate dockerfile
> 4. build docker image if needed or reload minimized image
> 5. run all tests via neurodocker reprozip (in parallel)
> 6. combine all reprozip packs into a single pack
> 7. generate minimized docker image using reprounzip and cache this image

Other changes:
- Migrate to CircleCI 2.0.

This is a work-in-progress. It would be great to have some feedback on what I have so far, namely dockerfile generation and the migration to circleci 2.0.

Parts of the code refer to my personal dockerhub repo (`kaczmarj/nipype`) but those will be changed when this PR is closer to complete. Also, the `master` neurodocker docker image is used now, but I will change this to `v0.3.2` once I release that version.

I am in the process of reworking neurodocker-reprozip's container-minimization functionality (ViDA-NYU/reprozip#274). I hope to finish that soon for this PR. I have already added a command to merge reprozip pack files (kaczmarj/neurodocker#69).

Just to note, I propose using `neurodebian:stretch-non-free` as the base image, instead of `ubuntu:xenial-20161213`.

Todo
-----
- [x] ~~generate dockerfiles using neurodocker~~
- [ ] check if dockerfile hash has changed or force update is set (need to determine this from PR)
    - checking is implemented, but forcing update is not.
- [ ] if force update, regenerate dockerfile
- [x] ~~build docker image if needed or reload minimized image~~
- [ ] run all tests via neurodocker reprozip (in parallel)
- [ ] combine all reprozip packs into a single pack
- [ ] generate minimized docker image using reprounzip and cache this image

### CODE REVIEW

**Comment by kaczmarj** [comment]
Please see https://circleci.com/gh/kaczmarj/nipype/182 for an example of the CircleCI run on my fork.

File: docker/generate_dockerfiles.sh [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+  --label maintainer="The nipype developers https://github.com/nipy/nipype" \
+  --spm version=12 matlab_version=R2017a \
+  --afni version=latest \
+  --freesurfer version=6.0.0 min=true \
+  --run 'echo "cHJpbnRmICJrcnp5c3p0b2YuZ29yZ29sZXdza2lAZ21haWwuY29tXG41MTcyXG4gKkN2dW12RVYzelRmZ1xuRlM1Si8yYzFhZ2c0RVxuIiA+IC9vcHQvZnJlZXN1cmZlci9saWNlbnNlLnR4dAo=" | base64 -d | sh' \
+  --install ants apt-utils bzip2 file fsl-core fsl-mni152-templates \
+            fusefat g++ git graphviz make ruby unzip xvfb \
+  --add-to-entrypoint "source /etc/fsl/fsl.sh" \
+  --env ANTSPATH='/usr/lib/ants' PATH='/usr/lib/ants:$PATH' \
+  --c3d version=1.0.0 \
```
Comment (satra) [inline docker/generate_dockerfiles.sh]: the latest version of convert3d might be available in neurodebian

**Comment by satra** [comment]
@kaczmarj - please make sure to merge with master, when sending an update.

**Comment by kaczmarj** [comment]
@satra - i'll be sure to do that. i am waiting on ViDA-NYU/reprozip#281. At the moment, `reprozip trace` returns status code 0 even if a traced command failed.

**Comment by kaczmarj** [comment]
In an effort to complete this PR in time for the release of version 0.14, minimizing containers with ReproZip will be done in a separate PR. The tests might have to be re-balanced in another PR.

Does it seem reasonable to use CircleCI's option to rebuild without cache to force a rebuild of the base Docker image?

Notes
-----

- The Dockerfile `ADD`/`COPY` instructions do not honor the current user, so the owner of the directories has to be manually changed to user neuro (https://github.com/moby/moby/issues/6119). Docker plans on changing this behavior by added a `--chown` flag to the `ADD`/`COPY` commands (https://github.com/moby/moby/pull/34263 and https://github.com/docker/cli/pull/467 for documentation). Looks like this is available in [Docker v17.09.0-ce-rc1](https://github.com/docker/docker-ce/releases/tag/v17.09.0-ce-rc1).
- CircleCI 2.0 does not have the environment variable `CIRCLE_TEST_REPORTS`. Use `store_artifacts` in `config.yml` instead (https://circleci.com/docs/2.0/configuration-reference/#store_artifacts).

File: .circleci/config.yml [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+            docker pull kaczmarj/nipype:base
+      - run:
+          name: Build main image (py36)
+          no_output_timeout: 60m
+          working_directory: /home/circleci/nipype
+          command: |
+            e=1 && for i in {1..5}; do
+              docker build \
+                --rm=false \
+                --tag kaczmarj/nipype:latest \
```
Comment (mgxd) [inline .circleci/config.yml]: nipype/nipype

File: .circleci/config.yml [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+            done && [ "$e" -eq "0" ]
+      - run:
+          name: Build main image (py27)
+          no_output_timeout: 60m
+          working_directory: /home/circleci/nipype
+          command: |
+            e=1 && for i in {1..5}; do
+              docker build \
+                --rm=false \
+                --tag kaczmarj/nipype:py27 \
```
Comment (mgxd) [inline .circleci/config.yml]: nipype/nipype

File: .circleci/config.yml [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+            chmod -R 777 "$WORKDIR"
+            bash /home/circleci/nipype/.circleci/tests.sh
+      - store_artifacts:
+          path: /home/circleci/work/tests
+      - run:
+          name: Save Docker images to workspace
+          no_output_timeout: 60m
+          command: |
+            if [ "$CIRCLE_NODE_INDEX" -eq "0" ]; then
+              docker save kaczmarj/nipype:base \
```
Comment (mgxd) [inline .circleci/config.yml]: nipype/nipype

File: .circleci/config.yml [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+          name: Load saved Docker images.
+          no_output_timeout: 60m
+          command: |
+            docker load < /tmp/docker/nipype-base-latest-py36-py27.tar
+      - run:
+          name: Push to DockerHub
+          no_output_timeout: 120m
+          command: |
+            echo "$DOCKER_PASS" | docker login -u "$DOCKER_USER" --password-stdin
+            docker push kaczmarj/nipype:base
```
Comment (mgxd) [inline .circleci/config.yml]: nipype/nipype

File: .circleci/config.yml [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+    jobs:
+      - compare_base_dockerfiles
+      - build_and_test:
+          requires:
+            - compare_base_dockerfiles
+      - deploy:
+          filters:
+            branches:
+              # TODO: change this to master after we are sure this works.
+              only: enh/circleci-neurodocker
```
Comment (mgxd) [inline .circleci/config.yml]: whoo!

File: .circleci/tests.sh [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+set -u         # Treat unset variables as an error when substituting.
+set -x         # Print command traces before executing command.
+
+if [ "${CIRCLE_NODE_TOTAL:-}" != "4" ]; then
+  echo "These tests were designed to be run at 4x parallelism."
+  exit 1
+fi
+
+# TODO: change this image name
+DOCKER_IMAGE="kaczmarj/nipype"
```
Comment (mgxd) [inline .circleci/tests.sh]: ping here

File: Dockerfile [hunk 8]
```diff
@@ -1,111 +1,233 @@
...
+    && conda clean -tipsy && sync
+
+#-------------------------
+# Create conda environment
+#-------------------------
+RUN conda create -y -q --name neuro \
+    && sync && conda clean -tipsy && sync \
+    && sed -i '$isource activate neuro' $ND_ENTRYPOINT
 
+COPY ["docker/files/run_builddocs.sh", "docker/files/run_examples.sh", "docker/files/run_pytests.sh", "nipype/external/fsl_imglob.py", "/usr/bin/"]
```
Comment (mgxd) [inline Dockerfile]: i don't feel to strongly about changing this, but maybe we can take advantage of `--chown` flag now

File: docker/generate_dockerfiles.sh [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+done
+
+
+# neurodocker version 0.3.1-19-g8d02eb4
+NEURODOCKER_IMAGE="kaczmarj/neurodocker@sha256:6b5f92f413b9710b7581e62293a8f74438b14ce7e4ab1ce68db2a09f7c64375a"
+
+# neurodebian:stretch-non-free pulled on November 3, 2017
+BASE_IMAGE="neurodebian@sha256:7590552afd0e7a481a33314724ae27f76ccedd05ffd7ac06ec38638872427b9b"
+
+NIPYPE_BASE_IMAGE="kaczmarj/nipype:base"
```
Comment (mgxd) [inline docker/generate_dockerfiles.sh]: nipype/nipype

File: .circleci/config.yml [hunk 10]
```diff
@@ -0,0 +1,190 @@
+version: 2
+jobs:
+
+  compare_base_dockerfiles:
+    docker:
+      - image: docker:17.09.0-ce-git
```
[the remark this answers is not in the document]
Comment (kaczmarj) [reply .circleci/config.yml]: that's a good point, although the change would be in the machine executor of `build_and_test`. i can change that to `circleci/classic:201710-02` (more info on [machine executors](https://circleci.com/docs/2.0/configuration-reference/#machine)).

File: .circleci/config.yml [hunk 11]
```diff
@@ -0,0 +1,190 @@
...
+      - run:
+          name: Prune base Dockerfile in preparation for cache check
+          working_directory: /home/circleci/nipype/docker
+          command: |
+            mkdir -p /tmp/docker
+            # Use the sha256 sum of the pruned Dockerfile as the cache key.
+            ash prune_dockerfile.sh Dockerfile.base > /tmp/docker/Dockerfile.base-pruned
+      - restore_cache:
+          # TODO: change this to 'master' after we are sure this works.
+          key: dftest-v5-enh/circleci-neurodocker-{{ checksum "/tmp/docker/Dockerfile.base-pruned" }}
```
[the remark this answers is not in the document]
Comment (kaczmarj) [reply .circleci/config.yml]: yes i will change this

File: .circleci/config.yml [hunk 12]
```diff
@@ -0,0 +1,190 @@
...
+            if [ "$CIRCLE_TAG" != "" ]; then
+              sed -i -E "s/(__version__ = )'[A-Za-z0-9.-]+'/\1'$CIRCLE_TAG'/" nipype/info.py
+            fi
+      - run:
+          name: Get base image (pull or build)
+          no_output_timeout: 60m
+          # TODO: remove `docker pull` once once caching works.
+          command: |
+            # bash /tmp/docker/get_base_image.sh
+            docker pull kaczmarj/nipype:base
```
[the remark this answers is not in the document]
Comment (kaczmarj) [reply .circleci/config.yml]: also has to be changed. this was in place to cut down on testing time when base docker image didn't change but the cache wasn't set yet

File: Dockerfile (hunk 8 above)
Comment (kaczmarj) [reply Dockerfile]: yes we can do that. although this would require an update to neurodocker. could this be done in a future pr?

Comment (mgxd) [reply Dockerfile]: yes

**Comment by satra** [comment]
i think this looks good - one question: since we are generating the dockerfile do we need to store it?

**Comment by kaczmarj** [comment]
The Dockerfiles are not generated during the testing. I was thinking that someone would generate the Dockerfiles if needed, commit the changes, and push. The testing would use those Dockerfiles.

Do you want CircleCI to generate the Dockerfiles?

**Comment by satra** [comment]
@kaczmarj - since the testing uses docker and has a script that generates the dockerfile, can't that be generated during testing rather than being in the repo?

**Comment by kaczmarj** [comment]
That makes sense. Should both the base and main Dockerfiles be generated in CircleCI? Or would you like to have the main Dockerfile in the repo?

**Comment by satra** [comment]
i think anything that can be generated should be generated otherwise it can go out of sync.

**Comment by kaczmarj** [comment]
@satra - dockerfiles are generated in CircleCI now, and I removed them from the repo.

I should also mention that If the base Dockerfile is ever reverted to a previous cached version, the CircleCI cache prefix (i.e., `'dockerfile-cache-v1'`) should be changed (perhaps incremented to `'dockerfile-cache-v2'`). Here are the two relevant lines [1](https://github.com/kaczmarj/nipype/blob/28547ccf8d3d9ddd068ba3d57ee8bc217f61de54/.circleci/config.yml#L21) and [2](https://github.com/kaczmarj/nipype/blob/28547ccf8d3d9ddd068ba3d57ee8bc217f61de54/.circleci/config.yml#L181
).

If the base Dockerfile is reverted to be the same (identical, not considering comments and empty lines) as a previous, cached base Dockerfile, the old cached Dockerfile will be found and the current base Docker image will be pulled. This is not desirable, but it happens because [CircleCI's caches are immutable](https://circleci.com/docs/2.0/caching/#clearing-cache).

**Comment by mgxd** [comment]
@kaczmarj don't forget to add yourself to the `.zenodo` file as well

**Pull request merged**